logbook-json/src/main/java/org/zalando/logbook/json/JsonHttpLogFormatter.java

```diff
@@ -1,92 +1,191 @@
+
+        StringBuilderWriter writer = new StringBuilderWriter(body.length() + 2048);
+
+        JsonGenerator generator = jsonFactory.createGenerator(writer);
+
+        try {
+            generator.writeStartObject();
+    
+            generator.writeFieldName("origin");
+            if(response.getOrigin() == Origin.LOCAL) {
+                generator.writeRawValue("\"local\"");
+            } else {
+                generator.writeRawValue("\"remote\"");
+            }
+            generator.writeFieldName("type");
+            generator.writeRawValue("\"response\"");
+            generator.writeStringField("correlation", correlationId);
+            generator.writeStringField("protocol", response.getProtocolVersion());
+            generator.writeNumberField("duration", correlation.getDuration().toMillis());
+            generator.writeNumberField("status", response.getStatus());
+    
+            writeHeaders(generator, response);
+            writeBody(response, body, generator);
+            
+            generator.writeEndObject();
+        } finally {
+            generator.close();
+        }
+        return writer.toString();
     }
 
-    @AllArgsConstructor
-    private static final class JsonBody {
-        String json;
+    protected void writeHeaders(JsonGenerator builder, HttpMessage httpMessage) throws IOException {
+        Map<String, List<String>> headers = httpMessage.getHeaders();
+        
+        if(!headers.isEmpty()) {
+            // implementation note:
+            // for some unclear reason, manually iterating over the headers
+            // while writing performs worse than letting Jackson do the job.
+            builder.writeObjectField("headers", headers);
+        }
+    }
+    
+    private void reconstruct(final HttpRequest request, JsonGenerator generator, StringBuilderWriter writer) throws IOException {
 
-        @JsonRawValue
-        @JsonValue
-        public String getJson() {
-            return json;
+        generator.writeRawValue("\"");
+
+        // write to underlying stream to avoid creating objects
+        
+        // first flush the json generator
+        generator.flush();
+
+        // then write to the underlying writer / builder
+        StringBuilder builder = writer.getBuilder();
+
+        final String scheme = request.getScheme();
+        builder.append(scheme);
+        builder.append("://"); // forward slash escaping is optional, so don't escape
+        
+        // host legal characters: Letters, Numbers 0-9 and Hyphen - none of which are escaped in JSON
+        builder.append(request.getHost()); 
+
+        final Optional<Integer> port = request.getPort();
+        if(port.isPresent() && isNotStandardPort(scheme, port.get())) {
+            builder.append(':').append(port.get());
         }
+        
+        JsonStringEncoder jsonStringEncoder = BufferRecyclers.getJsonStringEncoder();
+        jsonStringEncoder.quoteAsString(request.getPath(), builder); // escape
```

Comment:
The `JsonGenerator` doesn't give you this?

Reply:
It does, however I've tried to avoid escaping whenever possible. So reconstructUri is a bit ugly. Would be possible to make it a bit less low-level and see how that affects performance.